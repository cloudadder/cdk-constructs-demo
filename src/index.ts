import { Annotations, IAspect, Stack, Tags } from 'aws-cdk-lib';
import { CfnBucket } from 'aws-cdk-lib/aws-s3';
import { IConstruct } from 'constructs';


export interface CloudCostManagerProps {
  /**
   * The name of the customer that the stack is for.
   * e.g. ```'acme-co'```
   */
  readonly customerName: string;

  /**
 * The name of the environment.
 * e.g. ```'production'```
 */
  readonly envName: string;
}

export class CloudCostManager implements IAspect {
  private error!: string;
  private stack: Stack;
  private props: CloudCostManagerProps;

  constructor(stack: Stack, props: CloudCostManagerProps) {
    this.props = props;
    this.stack = stack;
    this.props = props;
    Tags.of(stack).add('cloud-cost-manager-customer-name', this.props.customerName);
    Tags.of(stack).add('cloud-cost-manager-env-name', this.props.envName);
    Tags.of(stack).add('cloud-cost-manager-version', '1.0.0');
  }

  visit(node: IConstruct): void {
    if (node instanceof CfnBucket ) {
      if (!node.intelligentTieringConfigurations) {
        this.error = 'Bucket requires intelligent tiering';
      } else {
        node.tags.setTag('cloud-cost-manager-check-intelligent-tiering', 'pass');
      }
    }
    if (this.error) {
      Annotations.of(this.stack).addError(this.error);
    } else {
      Annotations.of(this.stack).addInfo('CloudCostManager validation passed');
    }
  }
}


