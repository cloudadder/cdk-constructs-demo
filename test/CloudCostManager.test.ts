import { App, Aspects } from 'aws-cdk-lib';
import { Annotations, Template } from 'aws-cdk-lib/assertions';
import { CloudCostManager } from '../src';
import { ExistingBucketAfterAddingTiering, ExistingBucketWithErrors, TestStack, TestStackWithDatabase, TestStackWithErrors, TestStackWithoutAspect } from './TestStacks';

describe('Cloud Cost Manager', () => {
  test('confirm cloud cost management is implemented', () => {
    const app = new App();
    const stack = new TestStack(app, 'TestStack');

    Annotations.fromStack(stack).hasInfo('/TestStack',
      'CloudCostManager validation passed',
    );

    Template.fromStack(stack).hasResourceProperties('AWS::S3::Bucket', {
      Tags: [
        {
          Key: 'cloud-cost-manager:check:intelligent-tiering',
          Value: 'pass',
        },
        {
          Key: 'cloud-cost-manager:customer-name',
          Value: 'acme-co',
        },
        {
          Key: 'cloud-cost-manager:env-name',
          Value: 'staging',
        },
        {
          Key: 'cloud-cost-manager:version',
          Value: '1.0.0',
        },
      ],
    });
  });

  test('confirm cloud cost management reject stack', () => {
    const app = new App();
    var stack = new TestStackWithErrors(app, 'TestStack');

    Annotations.fromStack(stack).hasError('/TestStack',
      'Bucket requires intelligent tiering',
    );

    stack = new TestStackWithDatabase(app, 'TestStackWithDatabase');

    Annotations.fromStack(stack).hasError('/TestStackWithDatabase',
      'Do not use MSSQL Enterprise Edition, it is too expensive.',
    );

    Annotations.fromStack(stack).hasError('/TestStackWithDatabase',
      'Multi-AZ is not supported in Non Prodcution Environments.',
    );
  });

  test('confirm cloud cost management reject stack when using existing cfn template', () => {
    const app = new App();
    const stack = new ExistingBucketWithErrors(app, 'TestStack');

    Annotations.fromStack(stack).hasError('/TestStack',
      'Bucket requires intelligent tiering',
    );
  });

  test('confirm cloud cost management is implemented', () => {
    const app = new App();
    const stack = new ExistingBucketAfterAddingTiering(app, 'TestStack');

    Annotations.fromStack(stack).hasInfo('/TestStack',
      'CloudCostManager validation passed',
    );
  });

  test('confirm aspect works across app rather than just stack', () => {
    const app = new App();
    const stack = new TestStackWithoutAspect(app, 'TestStack');

    Aspects.of(app).add(new CloudCostManager(stack, {
      customerName: 'acme-co',
      envName: 'staging',
    }));

    Annotations.fromStack(stack).hasInfo('/TestStack',
      'CloudCostManager validation passed',
    );
  });

  test('Snapshot', () => {
    const app = new App();
    const stack = new TestStack(app, 'test');

    expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
  });
});