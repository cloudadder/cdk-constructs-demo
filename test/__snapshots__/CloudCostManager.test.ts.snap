// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Cost Manager Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "testbucketE6E05ABE": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "IntelligentTieringConfigurations": Array [
          Object {
            "Id": "test-tier",
            "Status": "Enabled",
            "Tierings": Array [],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager-check-intelligent-tiering",
            "Value": "pass",
          },
          Object {
            "Key": "cloud-cost-manager-customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager-env-name",
            "Value": "staging",
          },
          Object {
            "Key": "cloud-cost-manager-version",
            "Value": "1.0.0",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
